aiodocker/multiplexed.py

```diff
@@ -1,40 +1,78 @@
 import asyncio
 import struct
-
 import aiohttp
 
 from . import constants
 
+# aiohttp has no errors.py
+try:
+    import aiohttp.errors as errors
+except:
+    import aiohttp.client_exceptions as errors
+
 
 class MultiplexedResult:
     def __init__(self, response):
         self.response = response
 
     async def fetch(self):
-        while True:
-            try:
-                hdrlen = constants.STREAM_HEADER_SIZE_BYTES
-                header = await self.response.content.readexactly(hdrlen)
-                _, length = struct.unpack('>BxxxL', header)
-                if not length:
-                    continue
-                data = await self.response.content.readexactly(length)
-            except (aiohttp.ClientConnectionError,
-                    aiohttp.ServerDisconnectedError):
-                break
-            except asyncio.IncompleteReadError:
-                break
-            yield data.decode('utf8')
+        try:
+            while True:
+                try:
+                    hdrlen = constants.STREAM_HEADER_SIZE_BYTES
+                    header = await self.response.content.readexactly(hdrlen)
+                    _, length = struct.unpack('>BxxxL', header)
+                    if not length:
+                        continue
+                    data = await self.response.content.readexactly(length)
+                #except (aiohttp.errors.ClientDisconnectedError,
+                #        aiohttp.errors.ServerDisconnectedError):
+                except errors.ServerDisconnectedError:
+                    break
+                except asyncio.IncompleteReadError:
+                    break
+                yield data.decode('utf8')
+        finally:
+
+            await self.close()
+
+    async def fetch_raw(self):
+        try:
+            async for data in self.response.content.iter_chunked(1024):
+                yield data.decode('utf8')
+        finally:
+
+            await self.close()
 
     async def close(self):
-        await self.response.release()
 
+        if self.response.release():
+            await self.response.release()
 
-async def multiplexed_result(response, follow=False):
+
+async def multiplexed_result(response, follow=False, isTty=False):
     log_stream = MultiplexedResult(response)
-    if follow:
-        return log_stream
-    data = []
-    async for record in log_stream.fetch():
-        data.append(record)
-    return data
+
+    if isTty:
+        if follow:
+            return log_stream.fetch_raw()
+        else:
+            d = ''
+            async for l in log_stream.fetch_raw():
+                d = d + l
+
+            return d
+    else:
+        if follow:
+            return log_stream.fetch()
+        else:
+            d = ''
+            async for l in log_stream.fetch():
+                d = d + l
+
+            return d
```

Comment:
Okay, I think your main contribution is here: separation of `fetch` and `fetch_raw`.
String concatenation is generally considered a bad idea in Python due to performance reasons.
Could you rewrite those using `io.BytesIO` or `b''.join()`?